File: compute/runtime/cencori-shim.mjs
#!/usr/bin/env node
/**
 * Cencori Node shim — wraps a user's Node/TS agent with the Runtime Contract v2.
 *
 * Baked into the Node base image. The generic-node adapter (and JS frameworks
 * that expose an agent rather than a server) point START_COMMAND here and set
 * AGENT_ENTRYPOINT="<file>[:export]". Dependency-free (built-in http); Node
 * >=23.6 imports .ts entries directly via native type-stripping.
 *
 * Contract surface:
 *   GET  /_health              liveness + entry
 *   GET  /_manifest            capabilities (streaming, framework)
 *   POST /invoke               sync sugar — runs to completion, returns {output}
 *   POST /runs                 start an async run → {id, status}
 *   GET  /runs/:id             run state (status, output/error, suspend info)
 *   GET  /runs/:id/events      SSE event stream (?after=<seq> to resume)
 *   POST /runs/:id/cancel      request cancellation (best-effort via AbortSignal)
 *   POST /runs/:id/resume      resume a suspended run with {resume|input}
 *
 * Events are normalized across frameworks so the platform renders one timeline:
 *   run.started · message · token · tool.call · tool.result ·
 *   run.suspended · run.output · run.failed · run.completed
 *
 * Runs live in-memory (one agent per machine). A durable run store lands with
 * the platform; reconnecting clients replay from the ?after cursor meanwhile.
 */
import http from 'node:http';
import { randomUUID } from 'node:crypto';
import { pathToFileURL } from 'node:url';
import { resolve } from 'node:path';
import { createRequire } from 'node:module';

const PORT = Number(process.env.PORT || 8080);
const SPEC = process.env.AGENT_ENTRYPOINT || '';
const FRAMEWORK = process.env.FRAMEWORK || '';

let entry = null;

async function load() {
    if (entry) return entry;
    const [file, attr] = SPEC.split(':');
    if (!file) throw new Error('AGENT_ENTRYPOINT is not set (expected "<file>[:export]")');
    const mod = await import(pathToFileURL(resolve(process.cwd(), file)).href);
    entry = attr && mod[attr] != null ? mod[attr] : mod.default ?? mod.agent ?? mod;
    if (entry == null) throw new Error(`No usable export found in ${file}`);
    return entry;
}

/** Resolve a package from the AGENT's node_modules (not the shim's location). */
async function fromAgent(pkg) {
    const req = createRequire(pathToFileURL(resolve(process.cwd(), 'package.json')).href);
    return import(pathToFileURL(req.resolve(pkg)).href);
}

// ── Framework-aware invocation ─────────────────────────────────────────────

/** Sync call — runs to completion, returns the final output. */
async function callAgent(input) {
    const e = await load();
    if (FRAMEWORK === 'openai-agents') {
        const { run } = await fromAgent('@openai/agents');
        const result = await run(e, input);
        return result?.finalOutput ?? result;
    }
    if (FRAMEWORK === 'mastra' && e && typeof e.generate === 'function') {
        const result = await e.generate(input);
        return result?.text ?? result;
    }
    let out;
    if (e && typeof e.invoke === 'function') out = e.invoke(input);
    else if (typeof e === 'function') out = e(input);
    else if (e && typeof e.run === 'function') out = e.run(input);
    else if (e && typeof e.generate === 'function') out = e.generate(input);
    else throw new Error('AGENT_ENTRYPOINT is not callable and has no .invoke()/.run()');
    return await out;
}

/** LangGraph interrupts surface as `__interrupt__` on a streamed/returned chunk. */
function langgraphInterrupt(chunk) {
    const i = chunk && (chunk.__interrupt__ ?? chunk['__interrupt__']);
    if (i == null) return null;
    return { reason: 'interrupt', data: i };
}

/**
 * Stream an agent as normalized events. Yields any of:
 *   { type: 'message'|'token'|'tool.call'|'tool.result', data }
 *   { type: 'run.suspended', reason, data }   (terminal-for-now; resumable)
 *   { type: 'run.output', output }            (final result)
 * Frameworks without a stream method run sync and yield a single run.output.
 */
async function* streamAgent(input, signal) {
    const e = await load();

    // OpenAI Agents SDK — streaming run yields SDK events; finalOutput at the end.
    if (FRAMEWORK === 'openai-agents') {
        const { run } = await fromAgent('@openai/agents');
        const result = await run(e, input, { stream: true });
        if (result && typeof result[Symbol.asyncIterator] === 'function') {
            for await (const ev of result) {
                if (signal?.aborted) return;
                // Interruptions (approvals/handoffs) → suspended run.
                if (ev && (ev.type === 'run_item_stream_event') && ev.item?.type === 'tool_approval_item') {
                    yield { type: 'run.suspended', reason: 'approval', data: ev.item };
                    return;
                }
                yield { type: 'message', data: ev };
            }
        }
        yield { type: 'run.output', output: result?.finalOutput ?? result };
        return;
    }

    // LangGraph.js / LangChain / Mastra — native async-iterable .stream().
    if (e && typeof e.stream === 'function') {
        const stream = await e.stream(input, signal ? { signal } : undefined);
        let last;
        for await (const chunk of stream) {
            if (signal?.aborted) return;
            const suspend = langgraphInterrupt(chunk);
            if (suspend) {
                yield { type: 'run.suspended', reason: suspend.reason, data: suspend.data };
                return;
            }
            last = chunk;
            yield { type: 'message', data: chunk };
        }
        yield { type: 'run.output', output: last };
        return;
    }

    // No native streaming — run to completion, emit a single output event.
    yield { type: 'run.output', output: await callAgent(input) };
}

/** Resume a suspended run. LangGraph resumes by re-invoking with a Command. */
async function* resumeAgent(run, payload, signal) {
    const e = await load();
    if (e && typeof e.stream === 'function') {
        // LangGraph: `new Command({ resume })` — construct loosely to avoid a dep.
        const command = { resume: payload?.resume ?? payload?.input ?? payload };
        const stream = await e.stream(command, signal ? { signal } : undefined);
        let last;
        for await (const chunk of stream) {
            if (signal?.aborted) return;
            const suspend = langgraphInterrupt(chunk);
            if (suspend) {
                yield { type: 'run.suspended', reason: suspend.reason, data: suspend.data };
                return;
            }
            last = chunk;
            yield { type: 'message', data: chunk };
        }
        yield { type: 'run.output', output: last };
        return;
    }
    void run;
    throw new Error(`resume is not supported for framework "${FRAMEWORK || 'generic'}"`);
}

// ── In-memory run store + SSE fan-out ──────────────────────────────────────

/** @type {Map<string, Run>} */
const runs = new Map();

function createRun(input) {
    const run = {
        id: randomUUID(),
        status: 'running', // running | suspended | completed | failed | canceled
        input,
        output: undefined,
        error: undefined,
        suspend: undefined, // { reason, data } when status === 'suspended'
        events: [], // [{ seq, ts, type, ...payload }]
        seq: 0,
        subscribers: new Set(), // Set<http.ServerResponse>
        controller: new AbortController(),
        createdAt: Date.now(),
    };
    runs.set(run.id, run);
    return run;
}

function emit(run, event) {
    const record = { seq: ++run.seq, ts: Date.now(), ...event };
    run.events.push(record);
    const line = `id: ${record.seq}\nevent: ${record.type}\ndata: ${JSON.stringify(record)}\n\n`;
    for (const res of run.subscribers) res.write(line);
    return record;
}

const TERMINAL = new Set(['completed', 'failed', 'canceled']);

function finish(run, status, patch = {}) {
    if (TERMINAL.has(run.status)) return; // already settled — don't double-emit
    run.status = status;
    Object.assign(run, patch);
    emit(run, { type: 'run.completed', status, ...(patch.error ? { error: patch.error } : {}) });
    for (const res of run.subscribers) res.end();
    run.subscribers.clear();
}

/** Drive a run's async iterator, emitting normalized events until terminal. */
async function driveRun(run, iterator) {
    emit(run, { type: 'run.started', runId: run.id });
    try {
        for await (const ev of iterator) {
            if (run.controller.signal.aborted) {
                finish(run, 'canceled');
                return;
            }
            if (ev.type === 'run.suspended') {
                run.suspend = { reason: ev.reason, data: ev.data };
                emit(run, ev);
                run.status = 'suspended';
                // Suspended is not terminal — keep subscribers open for resume.
                for (const res of run.subscribers) res.write(`event: ping\ndata: {}\n\n`);
                return;
            }
            if (ev.type === 'run.output') {
                run.output = ev.output;
                emit(run, ev);
                finish(run, 'completed', { output: ev.output });
                return;
            }
            emit(run, ev);
        }
        // Iterator ended. Don't overwrite a run that cancel/suspend already settled
        // (streamAgent returns cleanly on abort, ending this loop normally).
        if (run.controller.signal.aborted) return finish(run, 'canceled');
        if (run.status === 'running') finish(run, 'completed', { output: run.output ?? null });
    } catch (err) {
        if (run.controller.signal.aborted) return finish(run, 'canceled');
        const error = { code: 'run_failed', message: String(err?.message ?? err) };
        run.error = error;
        emit(run, { type: 'run.failed', error });
        finish(run, 'failed', { error });
    }
}

function runView(run) {
    return {
        id: run.id,
        status: run.status,
        output: run.output,
        error: run.error,
        suspend: run.suspend,
        seq: run.seq,
    };
}

// ── HTTP ───────────────────────────────────────────────────────────────────

function send(res, code, body) {
    let payload;
    try {
        payload = JSON.stringify(body);
    } catch {
        payload = JSON.stringify({ output: String(body?.output ?? body) });
    }
    res.writeHead(code, { 'Content-Type': 'application/json' });
    res.end(payload);
}

function readJson(req) {
    return new Promise((resolveBody, reject) => {
        let raw = '';
        req.on('data', (c) => { raw += c; });
        req.on('end', () => {
            if (!raw) return resolveBody({});
            try { resolveBody(JSON.parse(raw)); } catch { reject(new Error('bad_json')); }
        });
        req.on('error', reject);
    });
}

function streamEvents(req, res, run, afterSeq) {
    res.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache, no-transform',
        Connection: 'keep-alive',
        'X-Accel-Buffering': 'no',
    });
    // Replay buffered events past the cursor (reconnect support).
    for (const ev of run.events) {
        if (ev.seq > afterSeq) res.write(`id: ${ev.seq}\nevent: ${ev.type}\ndata: ${JSON.stringify(ev)}\n\n`);
    }
    // Terminal already? close after the replay.
    if (run.status === 'completed' || run.status === 'failed' || run.status === 'canceled') {
        return res.end();
    }
    run.subscribers.add(res);
    const keepalive = setInterval(() => res.write(': keepalive\n\n'), 15000);
    req.on('close', () => {
        clearInterval(keepalive);
        run.subscribers.delete(res);
    });
}

const server = http.createServer(async (req, res) => {
    const url = new URL(req.url, 'http://localhost');
    const path = url.pathname;

    if (req.method === 'GET' && path === '/_health') return send(res, 200, { ok: true, entry: SPEC });
    if (req.method === 'GET' && path === '/_manifest') {
        return send(res, 200, { contract: 'v2', shim: 'generic-node', framework: FRAMEWORK || 'generic', streaming: true, runs: true });
    }

    // Sync sugar: run to completion.
    if (req.method === 'POST' && path === '/invoke') {
        let body;
        try { body = await readJson(req); } catch { return send(res, 400, { error: { code: 'bad_json', message: 'Invalid JSON body' } }); }
        try {
            return send(res, 200, { output: await callAgent(body.input) });
        } catch (err) {
            return send(res, 500, { error: { code: 'invoke_failed', message: String(err?.message ?? err) } });
        }
    }

    // Start an async run.
    if (req.method === 'POST' && path === '/runs') {
        let body;
        try { body = await readJson(req); } catch { return send(res, 400, { error: { code: 'bad_json', message: 'Invalid JSON body' } }); }
        const run = createRun(body.input);
        // Fire-and-forget; the client subscribes to /events for progress.
        driveRun(run, streamAgent(body.input, run.controller.signal)).catch((err) => {
            console.error('[cencori-shim] run crashed:', err);
        });
        return send(res, 202, runView(run));
    }

    // /runs/:id[/events|/cancel|/resume]
    const m = path.match(/^\/runs\/([^/]+)(\/events|\/cancel|\/resume)?$/);
    if (m) {
        const run = runs.get(m[1]);
        if (!run) return send(res, 404, { error: { code: 'run_not_found', message: m[1] } });
        const sub = m[2];

        if (!sub && req.method === 'GET') return send(res, 200, runView(run));

        if (sub === '/events' && req.method === 'GET') {
            const after = Number(url.searchParams.get('after') || '0') || 0;
            return streamEvents(req, res, run, after);
        }

        if (sub === '/cancel' && req.method === 'POST') {
            if (run.status === 'running' || run.status === 'suspended') {
                run.controller.abort();
                finish(run, 'canceled');
            }
            return send(res, 200, runView(run));
        }

        if (sub === '/resume' && req.method === 'POST') {
            if (run.status !== 'suspended') return send(res, 409, { error: { code: 'not_suspended', message: `run is ${run.status}` } });
            let body;
            try { body = await readJson(req); } catch { return send(res, 400, { error: { code: 'bad_json', message: 'Invalid JSON body' } }); }
            run.status = 'running';
            run.suspend = undefined;
            run.controller = new AbortController();
            driveRun(run, resumeAgent(run, body, run.controller.signal)).catch((err) => {
                console.error('[cencori-shim] resume crashed:', err);
            });
            return send(res, 202, runView(run));
        }

        return send(res, 405, { error: { code: 'method_not_allowed', message: `${req.method} ${path}` } });
    }

    send(res, 404, { error: { code: 'not_found', message: `${req.method} ${path}` } });
});

server.listen(PORT, '0.0.0.0', () => console.error(`[cencori-shim] node shim (contract v2) on :${PORT} — entry=${SPEC} framework=${FRAMEWORK || 'generic'}`));

File: compute/runtime/python/cencori_shim.py
"""
Cencori Python shim — wraps a user's Python agent with the Runtime Contract v2.

Baked into the Python base image (compute/runtime/python). The generic-python
adapter points the machine at this via START_COMMAND. It imports the agent named
by AGENT_ENTRYPOINT ("module:attr") and serves it.

Contract surface (parity with the Node shim):
  GET  /_health              liveness + entry
  GET  /_manifest            capabilities
  POST /invoke               sync sugar — runs to completion, returns {output}
  POST /runs                 start an async run → {id, status}
  GET  /runs/{id}            run state
  GET  /runs/{id}/events     SSE event stream (?after=<seq> to resume)
  POST /runs/{id}/cancel     request cancellation (task.cancel)
  POST /runs/{id}/resume     resume a suspended run with {resume|input}

Events are normalized across frameworks so the platform renders one timeline:
  run.started · message · run.suspended · run.output · run.failed · run.completed

Runs live in-memory (one agent per machine); reconnecting clients replay from
the ?after cursor. A durable run store lands with the platform.
"""
import os
import sys
import json
import time
import asyncio
import importlib
import uuid

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, StreamingResponse
import uvicorn

app = FastAPI()
_ENTRY = None
FRAMEWORK = os.environ.get("FRAMEWORK", "")
TERMINAL = {"completed", "failed", "canceled"}
_RUNS: dict = {}


def _load():
    spec = os.environ.get("AGENT_ENTRYPOINT")
    if not spec:
        raise RuntimeError("AGENT_ENTRYPOINT is not set (module:attr)")
    module_name, _, attr = spec.partition(":")
    sys.path.insert(0, os.getcwd())
    module = importlib.import_module(module_name)
    return getattr(module, attr) if attr else module


def _entry():
    global _ENTRY
    if _ENTRY is None:
        _ENTRY = _load()
    return _ENTRY


def _safe(obj):
    """Coerce a framework object into something JSON-serializable."""
    try:
        json.dumps(obj)
        return obj
    except (TypeError, ValueError):
        return str(obj)


def _interrupt(chunk):
    """LangGraph interrupts surface as a `__interrupt__` key on a streamed chunk."""
    if isinstance(chunk, dict) and "__interrupt__" in chunk:
        return chunk["__interrupt__"]
    return None


async def _call(entry, payload):
    # LangChain/LangGraph Runnables expose .invoke; plain functions are callable.
    if hasattr(entry, "invoke"):
        result = entry.invoke(payload)
    elif callable(entry):
        result = entry(payload)
    else:
        raise TypeError("AGENT_ENTRYPOINT is not callable and has no .invoke()")
    if asyncio.iscoroutine(result):
        result = await result
    return result


async def _stream_agent(entry, payload):
    """Yield normalized events. Frameworks without async streaming run to completion."""
    # OpenAI Agents SDK (Python) — streamed run yields SDK events; final_output at end.
    if FRAMEWORK == "openai-agents":
        try:
            from agents import Runner  # openai-agents package
        except Exception:
            Runner = None
        if Runner is not None:
            result = Runner.run_streamed(entry, payload)
            async for ev in result.stream_events():
                yield {"type": "message", "data": _safe(ev)}
            yield {"type": "run.output", "output": _safe(getattr(result, "final_output", None))}
            return

    # LangGraph / LangChain — native async streaming.
    if hasattr(entry, "astream"):
        last = None
        async for chunk in entry.astream(payload):
            sus = _interrupt(chunk)
            if sus is not None:
                yield {"type": "run.suspended", "reason": "interrupt", "data": _safe(sus)}
                return
            last = chunk
            yield {"type": "message", "data": _safe(chunk)}
        yield {"type": "run.output", "output": _safe(last)}
        return

    # No async streaming — run to completion, emit a single output.
    yield {"type": "run.output", "output": _safe(await _call(entry, payload))}


async def _resume_agent(entry, payload):
    """Resume a suspended run. LangGraph resumes via Command(resume=...)."""
    resume_val = payload.get("resume", payload.get("input", payload)) if isinstance(payload, dict) else payload
    try:
        from langgraph.types import Command
        command = Command(resume=resume_val)
    except Exception:
        command = {"resume": resume_val}
    if hasattr(entry, "astream"):
        last = None
        async for chunk in entry.astream(command):
            sus = _interrupt(chunk)
            if sus is not None:
                yield {"type": "run.suspended", "reason": "interrupt", "data": _safe(sus)}
                return
            last = chunk
            yield {"type": "message", "data": _safe(chunk)}
        yield {"type": "run.output", "output": _safe(last)}
        return
    raise RuntimeError(f'resume is not supported for framework "{FRAMEWORK or "generic"}"')


# ── In-memory run store + SSE fan-out ──────────────────────────────────────

def _create_run(payload):
    run = {
        "id": str(uuid.uuid4()),
        "status": "running",  # running | suspended | completed | failed | canceled
        "input": payload,
        "output": None,
        "error": None,
        "suspend": None,
        "events": [],
        "seq": 0,
        "subscribers": set(),  # set[asyncio.Queue]
        "cancel": asyncio.Event(),
        "task": None,
        "created_at": _now(),
    }
    _RUNS[run["id"]] = run
    return run


def _now():
    return int(time.time() * 1000)


def _emit(run, event):
    run["seq"] += 1
    rec = {"seq": run["seq"], "ts": _now(), **event}
    run["events"].append(rec)
    for q in list(run["subscribers"]):
        q.put_nowait(rec)
    return rec


def _finish(run, status, **patch):
    if run["status"] in TERMINAL:
        return  # already settled — don't double-emit
    run["status"] = status
    run.update(patch)
    tail = {"error": patch["error"]} if patch.get("error") else {}
    _emit(run, {"type": "run.completed", "status": status, **tail})
    for q in list(run["subscribers"]):
        q.put_nowait(None)  # sentinel → close the SSE stream
    run["subscribers"].clear()


async def _drive(run, agen):
    _emit(run, {"type": "run.started", "runId": run["id"]})
    try:
        async for ev in agen:
            if run["cancel"].is_set():
                _finish(run, "canceled")
                return
            t = ev.get("type")
            if t == "run.suspended":
                run["suspend"] = {"reason": ev["reason"], "data": ev["data"]}
                _emit(run, ev)
                run["status"] = "suspended"  # not terminal — resumable
                return
            if t == "run.output":
                run["output"] = ev["output"]
                _emit(run, ev)
                _finish(run, "completed", output=ev["output"])
                return
            _emit(run, ev)
        # Iterator ended. Don't overwrite a run cancel/suspend already settled.
        if run["cancel"].is_set():
            _finish(run, "canceled")
        elif run["status"] == "running":
            _finish(run, "completed", output=run.get("output"))
    except asyncio.CancelledError:
        _finish(run, "canceled")
        raise
    except Exception as exc:  # noqa: BLE001 — surface any framework error
        err = {"code": "run_failed", "message": str(exc)}
        run["error"] = err
        _emit(run, {"type": "run.failed", "error": err})
        _finish(run, "failed", error=err)


def _run_view(run):
    return {
        "id": run["id"],
        "status": run["status"],
        "output": run["output"],
        "error": run["error"],
        "suspend": run["suspend"],
        "seq": run["seq"],
    }


def _sse(rec):
    return f'id: {rec["seq"]}\nevent: {rec["type"]}\ndata: {json.dumps(rec)}\n\n'


# ── HTTP ───────────────────────────────────────────────────────────────────

@app.get("/_health")
async def health():
    return {"ok": True, "entry": os.environ.get("AGENT_ENTRYPOINT")}


@app.get("/_manifest")
async def manifest():
    return {"contract": "v2", "shim": "generic-python", "framework": FRAMEWORK or "generic", "streaming": True, "runs": True}


@app.post("/invoke")
async def invoke(req: Request):
    body = await req.json()
    try:
        output = await _call(_entry(), body.get("input"))
    except Exception as exc:  # noqa: BLE001
        return JSONResponse({"error": {"code": "invoke_failed", "message": str(exc)}}, status_code=500)
    return {"output": _safe(output)}


@app.post("/runs")
async def start_run(req: Request):
    body = await req.json()
    run = _create_run(body.get("input"))
    run["task"] = asyncio.create_task(_drive(run, _stream_agent(_entry(), body.get("input"))))
    return JSONResponse(_run_view(run), status_code=202)


@app.get("/runs/{run_id}")
async def get_run(run_id: str):
    run = _RUNS.get(run_id)
    if not run:
        return JSONResponse({"error": {"code": "run_not_found", "message": run_id}}, status_code=404)
    return _run_view(run)


@app.get("/runs/{run_id}/events")
async def run_events(run_id: str, after: int = 0):
    run = _RUNS.get(run_id)
    if not run:
        return JSONResponse({"error": {"code": "run_not_found", "message": run_id}}, status_code=404)

    async def gen():
        # Subscribe first, then replay — so events between replay and subscribe
        # aren't lost. De-dupe live records already covered by the replay.
        q: asyncio.Queue = asyncio.Queue()
        run["subscribers"].add(q)
        replayed = after
        try:
            for ev in list(run["events"]):
                if ev["seq"] > after:
                    replayed = max(replayed, ev["seq"])
                    yield _sse(ev)
            if run["status"] in TERMINAL:
                return
            while True:
                rec = await q.get()
                if rec is None:
                    break
                if rec["seq"] > replayed:
                    yield _sse(rec)
        finally:
            run["subscribers"].discard(q)

    return StreamingResponse(
        gen(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache, no-transform", "X-Accel-Buffering": "no"},
    )


@app.post("/runs/{run_id}/cancel")
async def cancel_run(run_id: str):
    run = _RUNS.get(run_id)
    if not run:
        return JSONResponse({"error": {"code": "run_not_found", "message": run_id}}, status_code=404)
    if run["status"] in ("running", "suspended"):
        run["cancel"].set()
        if run["task"] and not run["task"].done():
            run["task"].cancel()
        _finish(run, "canceled")
    return _run_view(run)


@app.post("/runs/{run_id}/resume")
async def resume_run(run_id: str, req: Request):
    run = _RUNS.get(run_id)
    if not run:
        return JSONResponse({"error": {"code": "run_not_found", "message": run_id}}, status_code=404)
    if run["status"] != "suspended":
        return JSONResponse({"error": {"code": "not_suspended", "message": f'run is {run["status"]}'}}, status_code=409)
    body = await req.json()
    run["status"] = "running"
    run["suspend"] = None
    run["cancel"] = asyncio.Event()
    run["task"] = asyncio.create_task(_drive(run, _resume_agent(_entry(), body)))
    return JSONResponse(_run_view(run), status_code=202)


if __name__ == "__main__":
    try:
        _ENTRY = _load()
        print(f"[cencori-shim] loaded {os.environ.get('AGENT_ENTRYPOINT')} (contract v2, framework={FRAMEWORK or 'generic'})", file=sys.stderr)
    except Exception as exc:  # noqa: BLE001 — defer failure to first call
        print(f"[cencori-shim] could not import AGENT_ENTRYPOINT yet: {exc}", file=sys.stderr)
    uvicorn.run(app, host="0.0.0.0", port=int(os.environ.get("PORT", "8080")))
